pkg/state/key.go

```diff
@@ -0,0 +1,162 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/kong/go-database-reconciler/pkg/utils"
+)
+
+const (
+	keyTableName = "key"
+)
+
+var keyTableSchema = &memdb.TableSchema{
+	Name: keyTableName,
+	Indexes: map[string]*memdb.IndexSchema{
+		"id": {
+			Name:    "id",
+			Unique:  true,
+			Indexer: &memdb.StringFieldIndex{Field: "ID"},
+		},
+		"name": {
+			Name:    "name",
+			Unique:  true,
+			Indexer: &memdb.StringFieldIndex{Field: "Name"},
+		},
+		all: allIndex,
+	},
+}
+
+// KeysCollection stores and indexes Kong Keys.
+type KeysCollection collection
+
+// Add adds a key to the collection.
+// key.ID should not be nil else an error is thrown.
+func (k *KeysCollection) Add(key Key) error {
+	if utils.Empty(key.ID) {
+		return errIDRequired
+	}
+	txn := k.db.Txn(true)
+	defer txn.Abort()
+	var searchBy []string
+	searchBy = append(searchBy, *key.ID)
+	if !utils.Empty(key.Name) {
+		searchBy = append(searchBy, *key.Name)
+	}
+	_, err := getKey(txn, searchBy...)
+	if err == nil {
+		return fmt.Errorf("inserting key %v: %w", key.Console(), ErrAlreadyExists)
+	} else if !errors.Is(err, ErrNotFound) {
+		return err
+	}
+	err = txn.Insert(keyTableName, &key)
+	if err != nil {
+		return err
+	}
+	txn.Commit()
+	return nil
+}
+
+func getKey(txn *memdb.Txn, IDs ...string) (*Key, error) {
+	for _, id := range IDs {
+		res, err := multiIndexLookupUsingTxn(txn, keyTableName,
+			[]string{"name", "id"}, id)
+		if errors.Is(err, ErrNotFound) {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		key, ok := res.(*Key)
+		if !ok {
+			panic(unexpectedType)
+		}
+		return &Key{Key: *key.DeepCopy()}, nil
+	}
+	return nil, ErrNotFound
+}
+
+// Get gets a key by name or ID.
+func (k *KeysCollection) Get(nameOrID string) (*Key, error) {
+	if nameOrID == "" {
+		return nil, errIDRequired
+	}
+	txn := k.db.Txn(false)
+	defer txn.Abort()
+	key, err := getKey(txn, nameOrID)
```

Comment:
Why not return return value from getKey?
Checking error type again seems redundant 😅